crates/hir-def/src/nameres/path_resolution.rs

```diff
@@ -42,6 +42,9 @@ pub(super) enum ReachedFixedPoint {
 pub(super) struct ResolvePathResult {
     pub(super) resolved_def: PerNs,
     pub(super) segment_index: Option<usize>,
+    /// When resolving `Enum::Variant`, this points at `Enum`. If `Variant` is in scope and appears
+    /// alone this is `None`.
+    pub(super) enum_segment: Option<usize>,
```

Comment:
Ah this generally doesn't handle self/crate etc path prefixes

Reply:
Either way I encoded the `enum_segment` part differently while experimenting/investigating, its not tracked in a bool-ish field (same field as the differing crate part), as the actual index can be inferred instead (it's either the second to last, or `None`). Imo this looks a bit cleaner as it prevents encoding more states than actually occur, if you are on board feel free to cherry pick this.

Reply:
`self` is not checked for generics because it loses it generics when lowered to HIR, so forbidding generics for them will have to be a parse/parse validation error.

I'll cherry pick your commit.